core/embed/trezorhal/flash.h

```diff
@@ -22,101 +22,47 @@
-#define FLASH_SECTOR_STORAGE_1 4
-#define FLASH_SECTOR_STORAGE_2 16
-#elif defined TREZOR_MODEL_1
-#define FLASH_SECTOR_STORAGE_1 2
-#define FLASH_SECTOR_STORAGE_2 3
-#else
-#error Unknown Trezor model
-#endif
-
-#define FLASH_SECTOR_BOOTLOADER 5
-
-#define FLASH_SECTOR_FIRMWARE_START 6
-//                                           7
-//                                           8
-//                                           9
-//                                          10
-#define FLASH_SECTOR_FIRMWARE_END 11
-
-#define FLASH_SECTOR_UNUSED_START 12
-//                                          13
-//                                          14
-#define FLASH_SECTOR_UNUSED_END 15
+#define FLASH_OTP_NUM_BLOCKS 16
+#define FLASH_OTP_BLOCK_SIZE 32
 
-#define FLASH_SECTOR_FIRMWARE_EXTRA_START 17
-//                                          18
-//                                          19
-//                                          20
-//                                          21
-//                                          22
-#define FLASH_SECTOR_FIRMWARE_EXTRA_END 23
+/**
+ * Flash driver interface is designed to abstract away differences between
+ * various MCUs used in Trezor devices.
+ *
+ * Generally, flash memory is divided into sectors. On different MCUs, sectors
+ * may have different sizes, and therefore, different number of sectors are used
+ * for a given purpose. For example, on STM32F4, the sectors are relatively
+ * large so we use single sector for Storage. On STM32U5, the sectors are
+ * smaller, so we use multiple sectors for the Storage. Storage implementation
+ * should not care about this, and should use flash_area_t interface to access
+ * the flash memory.
+ *
+ * flash_area_t represents a location in flash memory. It may be contiguous, or
+ * it may be composed of multiple non-contiguous subareas.
+ *
+ * flash_subarea_t represents a contiguous area in flash memory, specified by
+ * first_sector and num_sectors.
+ */
 
-#define BOOTLOADER_SECTORS_COUNT (1)
-#define STORAGE_SECTORS_COUNT (2)
-#define FIRMWARE_SECTORS_COUNT (6 + 7)
+#include "flash_common.h"
 
-extern const uint8_t STORAGE_SECTORS[STORAGE_SECTORS_COUNT];
-extern const uint8_t FIRMWARE_SECTORS[FIRMWARE_SECTORS_COUNT];
+void flash_init(void);
 
-// note: FLASH_SR_RDERR is STM32F42xxx and STM32F43xxx specific (STM32F427)
-// (reference RM0090 section 3.7.5)
-#if !defined STM32F427xx && !defined STM32F429xx
-#define FLASH_SR_RDERR 0
-#endif
+secbool flash_write_byte(uint16_t sector, uint32_t offset, uint8_t data);
 
-#define FLASH_STATUS_ALL_FLAGS                                            \
-  (FLASH_SR_RDERR | FLASH_SR_PGSERR | FLASH_SR_PGPERR | FLASH_SR_PGAERR | \
-   FLASH_SR_WRPERR | FLASH_SR_SOP | FLASH_SR_EOP)
+secbool flash_write_word(uint16_t sector, uint32_t offset, uint32_t data);
 
-void flash_init(void);
+secbool flash_area_write_byte(const flash_area_t *area, uint32_t offset,
+                              uint8_t data);
```

Comment:
Please remove `flash_area_write_byte()`. Already declared in `flash_common.h`.

Reply:
https://github.com/trezor/trezor-firmware/commit/626e6d0954357a79c03f7e8f80a947f9414f78f7